scripts/core/kernel.yml

```diff
@@ -358,6 +358,16 @@ params:
 returns:
     - $X_RESULT_ERROR_INVALID_KERNEL_ARGUMENT_INDEX
 --- #--------------------------------------------------------------------------
+type: struct
+desc: "Properties for for $xKernelSetArgMemObj."
```

Comment:
L0 doesnt set these memory properties per kernel, but per object. Do we have that functionality in CUDA or other adapters?

Reply:
@jandres742 AFAIK L0 plugin uses access mode to get a proper memory handle (per object) to set it as exact kernel argument (per kernel). Other plugins now just ignore this property.

Reply:
The closes L0 has is:
https://spec.oneapi.io/level-zero/latest/core/api.html?highlight=mostly#_CPPv418ze_memory_advice_t

but not WRITE_ONLY, READ_ONLY, READ_WRITE.

>AFAIK L0 plugin is intended to use access mode to get a proper memory handle (per object) to set it as exact kernel argument (per kernel).

could you elaborate on that? How the L0 plugin gets the memory handle based on access mode?


```
enumerator ZE_MEMORY_ADVICE_SET_READ_MOSTLY
hint that memory will be read from frequently and written to rarely

enumerator ZE_MEMORY_ADVICE_CLEAR_READ_MOSTLY
removes the affect of [ZE_MEMORY_ADVICE_SET_READ_MOSTLY](https://spec.oneapi.io/level-zero/latest/core/api.html?highlight=mostly#ze__api_8h_1a54593bb72c476f6c3ac90909ae126cdfaf2a7ce987fb033859d05bc5dd812b1f5)

enumerator ZE_MEMORY_ADVICE_SET_PREFERRED_LOCATION
hint that the preferred memory location is the specified device

enumerator ZE_MEMORY_ADVICE_CLEAR_PREFERRED_LOCATION
removes the affect of [ZE_MEMORY_ADVICE_SET_PREFERRED_LOCATION](https://spec.oneapi.io/level-zero/latest/core/api.html?highlight=mostly#ze__api_8h_1a54593bb72c476f6c3ac90909ae126cdfaf56067d28d73977bb30cc6a35d89520c)

enumerator ZE_MEMORY_ADVICE_SET_NON_ATOMIC_MOSTLY
hints that memory will mostly be accessed non-atomically

enumerator ZE_MEMORY_ADVICE_CLEAR_NON_ATOMIC_MOSTLY
removes the affect of [ZE_MEMORY_ADVICE_SET_NON_ATOMIC_MOSTLY](https://spec.oneapi.io/level-zero/latest/core/api.html?highlight=mostly#ze__api_8h_1a54593bb72c476f6c3ac90909ae126cdfa62a34e1d7d43897cd3d9318474ba88fd)

enumerator ZE_MEMORY_ADVICE_BIAS_CACHED
hints that memory should be cached

enumerator ZE_MEMORY_ADVICE_BIAS_UNCACHED
hints that memory should be not be cached

enumerator ZE_MEMORY_ADVICE_FORCE_UINT32[¶](https://spec.oneapi.io/level-zero/latest/core/api.html?highlight=mostly#_CPPv4N18ze_memory_advice_t29ZE_MEMORY_ADVICE_FORCE_UINT32E)
```

Reply:
clarified with @smaslov-intel  and yes, we are actually already implementing this in L0 adapter.

@kbenzie : about the advantages: if the argument is only R, we avoid having extra copies needed to update the argument if it were RW or W, so there's actually advantages on doing this. @smaslov-intel would correct me.